packages/contracts/contracts/split/SplitAtomicV0.sol

```diff
@@ -0,0 +1,254 @@
+     * Calling a `nonReentrant` function from another `nonReentrant`
+     * function is not supported. It is possible to prevent this from happening
+     * by making the `nonReentrant` function external, and making it call a
+     * `private` function that does the actual work.
+     */
+    modifier nonReentrant() {
+        _nonReentrantBefore();
+        _;
+        _nonReentrantAfter();
+    }
+
+    /**
+     * @notice receive function splits received funds according to the
+     * configured `splits`.
+     * Reverts if contract is not yet initialized.
+     * @dev This function automatically splits funds when the native token of a
+     * blockchain is sent to the contract. It is important to note that this
+     * function uses an unspecified amount of gas, and therefore sending funds
+     * to this contract via the deprecated `transfer` method is not supported.
+     * @dev This function relies on being non-reentrant for security.
+     */
+    receive() external payable nonReentrant {
+        // split received funds
+        // @dev reverts if not already initialized
+        _splitETH(msg.value);
+    }
+
+    /**
+     * @notice Initializes contract with the provided `splits`.
+     * @dev This function should be called atomically, immediately after
+     * deployment.
+     */
+    function initialize(Split[] calldata splits) external {
+        // initialize reentrancy guard
+        // @dev this also verifies not already initialized
+        _reentrancyGuardInit();
+        // validate splits
+        uint256 totalBasisPoints = 0;
+        uint256 splitsLength = splits.length;
+        // @dev splits length implicitly checked to be > 0 via totalBasisPoints
+        // check after loop
+        for (uint256 i; i < splitsLength; ) {
+            Split memory split = splits[i];
+            uint256 bps = split.basisPoints;
+            require(bps > 0 && bps <= 10_000, "Invalid basis points");
+            // push the splits to the storage array
+            _splits.push(split);
+            // track total basis points for totals validation after loop
+            // @dev overflow checked automatically in solidity 0.8
+            totalBasisPoints += bps;
+            // @dev efficient unchecked increment
+            unchecked {
+                ++i;
+            }
+        }
+        require(totalBasisPoints == 10_000, "Invalid total basis points");
+        // emit initialized event
+        emit Initialized();
+    }
+
+    /**
+     * @notice Drains the contract's balance to the configured `splits`.
+     * Reverts if not initialized.
+     * @dev This function is useful for draining the contract's balance to the
+     * configured `splits` in the event that the contract receives funds via
+     * a force-send (e.g. `SELFDESTRUCT` or `SENDALL`) operation.
+     * @dev This function relies on being non-reentrant for security.
+     */
+    function drainETH() external nonReentrant {
```

Comment:
Should we emit an event here and `drainERC20`?

Reply:
was considering that, but I can't see how it would affect our indexing, so was thinking less gas is better here? idk, what do you think?

Reply:
that makes sense 🙏 was thinking it might be helpful if we want to track/monitor those calls but that might just be noise?

Reply:
added here: [1049411](https://github.com/ArtBlocks/artblocks-contracts/pull/1268/commits/10494119e20b21b9d6b7fe689d1f45f762c417f6)